src/snowflake/cli/_plugins/snowpark/snowpark_project_paths.py

```diff
@@ -56,18 +74,126 @@ def snowflake_requirements(self) -> SecurePath:
+
+    project_root: Path
+    path: Path
+    dest: str | None = None
+
+    def __init__(
+        self, project_root: Path, path: Path, dest: Optional[str] = None
+    ) -> None:
+        self.project_root = project_root
+        self.path = path
+        self.dest = dest
+        if self.dest and not self._is_dest_a_file() and not self.dest.endswith("/"):
+            self.dest = self.dest + "/"
+
     @property
-    def dependencies(self) -> Path:
-        return self.path_relative_to_root(Path("dependencies.zip"))
+    def _artefact_name(self) -> str:
+        if glob.has_magic(str(self.path)):
+            last_part = None
+            for part in self.path.parts:
+                if glob.has_magic(part):
+                    break
+                else:
+                    last_part = part
+            if not last_part:
+                last_part = os.path.commonpath(
+                    [str(self.path), str(self.path.absolute())]
+                )
+                raise ValueError("Doopa")
+            # before_wildcard = str(self.path).split("*")[0]
+            # last_part = Path(before_wildcard).absolute().parts[-1]
+            return last_part + ".zip"
+        if (self.project_root / self.path).is_dir():
+            return self.path.stem + ".zip"
+        if (self.project_root / self.path).is_file() and self._is_dest_a_file():
+            return Path(self.dest).name  # type: ignore
+        return self.path.name
+
+    @property
+    def post_build_path(self) -> Path:
+        """
+        Returns post-build artefact path. Directories are mapped to corresponding .zip files.
+        """
+        deploy_root = self.deploy_root()
+        path = (
+            self._path_until_asterisk()
+            if glob.has_magic(str(self.path))
+            else self.path.parent
+        )
+        if self._is_dest_a_file():
+            return deploy_root / self.dest  # type: ignore
+        return deploy_root / (self.dest or path) / self._artefact_name
+
+    def upload_path(self, stage: FQN | str | None) -> str:
+        """
+        Path on stage to which the artefact should be uploaded.
+        """
+        stage = stage or DEPLOYMENT_STAGE
+        if isinstance(stage, str):
+            stage = FQN.from_stage(stage).using_context()
+
+        stage_path = PurePosixPath(f"@{stage}")
+        if self.dest:
+            stage_path /= (
+                PurePosixPath(self.dest).parent if self._is_dest_a_file() else self.dest
+            )
+        else:
+            stage_path /= (
+                self._path_until_asterisk()
+                if glob.has_magic(str(self.path))
+                else PurePosixPath(self.path).parent
+            )
+
+        return str(stage_path) + "/"
+
+    def import_path(self, stage: FQN | str | None) -> str:
+        """Path for UDF/sproc imports clause."""
+        return self.upload_path(stage) + self._artefact_name
+
+    def deploy_root(self) -> Path:
+        return self.project_root / "output"
```

Comment:
this could cause conflicts with other domains, since native apps uses output/deploy by default. We make ours configurable as well through snowflake.yml. We should probably align the choice of deploy root, and make it per-entity so that multiple entities, even across different domains, don't negatively interfere with each other. We should probably do that as a follow-up PR very quickly.